File: python/aim_node/gateway_v2/publish.py
from __future__ import annotations

from dataclasses import asdict, dataclass
from typing import Any, Literal

import httpx

from .contracts import CLIENT_METHODS, GatewaySurface, reject_payload_contract_fields


PUBLISH_CLIENT_METHOD = CLIENT_METHODS[GatewaySurface.PUBLISH]
PublishGovernanceState = Literal[
    "pending_review",
    "review_required",
    "abuse_blocked",
    "trust_requirement_unmet",
    "publish_denied",
    "published",
]


@dataclass(frozen=True)
class SignedReference:
    ref: str
    signer_identity: str
    signature_state: str
    verification_state: str
    artifact_hash: str | None = None
    expires_at: str | None = None


@dataclass(frozen=True)
class ListingMetadata:
    title: str
    summary: str
    tags: list[str] | None = None
    category: str | None = None


@dataclass(frozen=True)
class ListingPricing:
    model: str
    currency: str
    unit_amount_cents: int
    billing_unit: str


@dataclass(frozen=True)
class ListingSamplePolicy:
    preview_refs: list[SignedReference] | None = None
    summary: str | None = None
    row_count: int | None = None
    column_names: list[str] | None = None


@dataclass(frozen=True)
class LicenseTerms:
    terms_ref: SignedReference
    commercial_use: bool | None = None
    retention_days: int | None = None


@dataclass(frozen=True)
class PublishRequest:
    metadata: dict[str, Any]
    seller_id: str
    listing_metadata: ListingMetadata
    delivery_modes: list[str]
    pricing: ListingPricing
    license_terms: LicenseTerms
    sample_policy: ListingSamplePolicy | None = None
    trust_artifact_refs: list[SignedReference] | None = None
    seller_verification_ref: SignedReference | None = None
    listing_review_policy: str | None = None
    abuse_signals: list[str] | None = None


@dataclass(frozen=True)
class PatchListingRequest:
    metadata: dict[str, Any]
    listing_version_id: str | None = None
    listing_metadata: ListingMetadata | None = None
    delivery_modes: list[str] | None = None
    pricing: ListingPricing | None = None
    sample_policy: ListingSamplePolicy | None = None
    license_terms: LicenseTerms | None = None
    trust_artifact_refs: list[SignedReference] | None = None
    seller_verification_ref: SignedReference | None = None
    listing_review_policy: str | None = None
    abuse_signals: list[str] | None = None


@dataclass(frozen=True)
class PublishedListing:
    listing_id: str
    listing_version_id: str
    seller_id: str
    state: PublishGovernanceState
    discoverable: bool
    quote_eligible: bool
    listing_metadata: ListingMetadata
    delivery_modes: list[str]
    pricing: ListingPricing
    sample_policy: ListingSamplePolicy
    license_terms: LicenseTerms
    trust_artifact_refs: list[SignedReference]
    seller_verification_ref: SignedReference | None = None
    listing_review_policy: str | None = None
    governance_events: list[str] | None = None
    reason_code: str | None = None


class LocalPublishDraftCache:
    def __init__(self) -> None:
        self._drafts: dict[str, PublishRequest] = {}

    def save_draft(self, key: str, draft: PublishRequest) -> None:
        assert_publish_request_has_no_payload(draft)
        self._drafts[key] = draft

    def get_draft(self, key: str) -> PublishRequest | None:
        return self._drafts.get(key)

    def delete_draft(self, key: str) -> bool:
        return self._drafts.pop(key, None) is not None


class GatewayPublishClient:
    def __init__(
        self,
        base_url: str,
        api_key: str | None = None,
        client: httpx.Client | None = None,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.client = client or httpx.Client()
        self.draft_cache = LocalPublishDraftCache()

    def publish(self, request: PublishRequest) -> PublishedListing:
        assert_publish_request_has_no_payload(request)
        response = self.client.post(
            f"{self.base_url}/v1/gateway/publish",
            json=_drop_none(asdict(request)),
            headers=self._headers(request.metadata),
        )
        response.raise_for_status()
        return _parse_published_listing(response.json())

    def patch_listing(
        self,
        listing_id: str,
        request: PatchListingRequest,
        if_match: str | None = None,
    ) -> PublishedListing:
        assert_no_payload_deep(asdict(request))
        headers = self._headers(request.metadata)
        if if_match:
            headers["if-match"] = if_match
        response = self.client.patch(
            f"{self.base_url}/v1/gateway/listings/{listing_id}",
            json=_drop_none(asdict(request)),
            headers=headers,
        )
        response.raise_for_status()
        return _parse_published_listing(response.json())

    def _headers(self, metadata: dict[str, Any]) -> dict[str, str]:
        headers = {
            "content-type": "application/json",
            "x-request-id": str(_pick(metadata, "requestId", "request_id")),
        }
        idempotency_key = _pick(metadata, "idempotencyKey", "idempotency_key")
        if idempotency_key:
            headers["idempotency-key"] = str(idempotency_key)
        if self.api_key:
            headers["authorization"] = f"Bearer {self.api_key}"
        return headers


def assert_publish_request_has_no_payload(request: PublishRequest) -> None:
    assert_no_payload_deep(asdict(request))


def assert_no_payload_deep(value: Any) -> None:
    if isinstance(value, dict):
        reject_payload_contract_fields(value.keys())
        for child in value.values():
            assert_no_payload_deep(child)
    elif isinstance(value, list | tuple):
        for child in value:
            assert_no_payload_deep(child)


def _parse_published_listing(data: dict[str, Any]) -> PublishedListing:
    return PublishedListing(
        listing_id=str(_pick_required(data, "listing_id", "listingId")),
        listing_version_id=str(_pick_required(data, "listing_version_id", "listingVersionId")),
        seller_id=str(_pick_required(data, "seller_id", "sellerId")),
        state=str(_pick_required(data, "state")),
        discoverable=bool(_pick_required(data, "discoverable")),
        quote_eligible=bool(_pick_required(data, "quote_eligible", "quoteEligible")),
        listing_metadata=_parse_listing_metadata(_pick_required(data, "listing_metadata", "listingMetadata")),
        delivery_modes=list(_pick_required(data, "delivery_modes", "deliveryModes")),
        pricing=_parse_pricing(_pick_required(data, "pricing")),
        sample_policy=_parse_sample_policy(_pick_required(data, "sample_policy", "samplePolicy")),
        license_terms=_parse_license_terms(_pick_required(data, "license_terms", "licenseTerms")),
        trust_artifact_refs=[
            _parse_signed_reference(item) for item in _pick(data, "trust_artifact_refs", "trustArtifactRefs") or []
        ],
        seller_verification_ref=_parse_optional_signed_reference(
            _pick(data, "seller_verification_ref", "sellerVerificationRef")
        ),
        listing_review_policy=_pick(data, "listing_review_policy", "listingReviewPolicy"),
        governance_events=_pick(data, "governance_events", "governanceEvents"),
        reason_code=_pick(data, "reason_code", "reasonCode"),
    )


def _parse_signed_reference(data: dict[str, Any]) -> SignedReference:
    return SignedReference(
        ref=data["ref"],
        artifact_hash=_pick(data, "artifact_hash", "artifactHash"),
        signer_identity=str(_pick_required(data, "signer_identity", "signerIdentity")),
        signature_state=str(_pick_required(data, "signature_state", "signatureState")),
        verification_state=str(_pick_required(data, "verification_state", "verificationState")),
        expires_at=_pick(data, "expires_at", "expiresAt"),
    )


def _parse_optional_signed_reference(data: dict[str, Any] | None) -> SignedReference | None:
    return _parse_signed_reference(data) if data else None


def _parse_listing_metadata(data: dict[str, Any]) -> ListingMetadata:
    return ListingMetadata(
        title=data["title"],
        summary=data["summary"],
        tags=data.get("tags"),
        category=data.get("category"),
    )


def _parse_pricing(data: dict[str, Any]) -> ListingPricing:
    return ListingPricing(
        model=data["model"],
        currency=data["currency"],
        unit_amount_cents=int(_pick_required(data, "unit_amount_cents", "unitAmountCents")),
        billing_unit=str(_pick_required(data, "billing_unit", "billingUnit")),
    )


def _parse_sample_policy(data: dict[str, Any]) -> ListingSamplePolicy:
    return ListingSamplePolicy(
        preview_refs=[_parse_signed_reference(item) for item in _pick(data, "preview_refs", "previewRefs") or []],
        summary=data.get("summary"),
        row_count=_pick(data, "row_count", "rowCount"),
        column_names=_pick(data, "column_names", "columnNames"),
    )


def _parse_license_terms(data: dict[str, Any]) -> LicenseTerms:
    return LicenseTerms(
        terms_ref=_parse_signed_reference(_pick_required(data, "terms_ref", "termsRef")),
        commercial_use=_pick(data, "commercial_use", "commercialUse"),
        retention_days=_pick(data, "retention_days", "retentionDays"),
    )


def _pick(data: dict[str, Any], *keys: str) -> Any:
    for key in keys:
        if key in data:
            return data[key]
    return None


def _pick_required(data: dict[str, Any], *keys: str) -> Any:
    value = _pick(data, *keys)
    if value is None:
        raise KeyError(keys[0])
    return value


def _drop_none(data: dict[str, Any]) -> dict[str, Any]:
    return {
        key: _drop_none(value) if isinstance(value, dict) else value
        for key, value in data.items()
        if value is not None
    }

File: src/gateway_v2/publish.ts
import { assertNoPayloadFields, clientMethods } from "./client_contracts";
import type { OpaqueId, SharedRequestMetadata } from "./client_contracts";

export const publishClientMethod = clientMethods.publish;

export type PublishGovernanceState =
  | "pending_review"
  | "review_required"
  | "abuse_blocked"
  | "trust_requirement_unmet"
  | "publish_denied"
  | "published";

export interface SignedReference {
  ref: string;
  artifactHash?: string;
  signerIdentity: OpaqueId;
  signatureState: string;
  verificationState: string;
  expiresAt?: string;
}

export interface ListingMetadata {
  title: string;
  summary: string;
  tags?: string[];
  category?: string;
}

export interface ListingPricing {
  model: string;
  currency: string;
  unitAmountCents: number;
  billingUnit: string;
}

export interface ListingSamplePolicy {
  previewRefs?: SignedReference[];
  summary?: string;
  rowCount?: number;
  columnNames?: string[];
}

export interface LicenseTerms {
  termsRef: SignedReference;
  commercialUse?: boolean;
  retentionDays?: number;
}

export interface PublishRequest {
  metadata: SharedRequestMetadata;
  sellerId: OpaqueId;
  listingMetadata: ListingMetadata;
  deliveryModes: string[];
  pricing: ListingPricing;
  samplePolicy?: ListingSamplePolicy;
  licenseTerms: LicenseTerms;
  trustArtifactRefs?: SignedReference[];
  sellerVerificationRef?: SignedReference;
  listingReviewPolicy?: string;
  abuseSignals?: string[];
}

export interface PatchListingRequest {
  metadata: SharedRequestMetadata;
  listingVersionId?: OpaqueId;
  listingMetadata?: ListingMetadata;
  deliveryModes?: string[];
  pricing?: ListingPricing;
  samplePolicy?: ListingSamplePolicy;
  licenseTerms?: LicenseTerms;
  trustArtifactRefs?: SignedReference[];
  sellerVerificationRef?: SignedReference;
  listingReviewPolicy?: string;
  abuseSignals?: string[];
}

export interface PublishedListing {
  listingId: OpaqueId;
  listingVersionId: OpaqueId;
  sellerId: OpaqueId;
  state: PublishGovernanceState;
  discoverable: boolean;
  quoteEligible: boolean;
  listingMetadata: ListingMetadata;
  deliveryModes: string[];
  pricing: ListingPricing;
  samplePolicy: ListingSamplePolicy;
  licenseTerms: LicenseTerms;
  trustArtifactRefs: SignedReference[];
  sellerVerificationRef?: SignedReference;
  listingReviewPolicy?: string;
  governanceEvents?: string[];
  reasonCode?: string;
}

export interface GatewayPublishClientOptions {
  baseUrl: string;
  apiKey?: string;
  fetch?: typeof fetch;
}

export class LocalPublishDraftCache {
  private readonly drafts = new Map<string, PublishRequest>();

  saveDraft(key: string, draft: PublishRequest): void {
    assertPublishRequestHasNoPayload(draft);
    this.drafts.set(key, structuredClone(draft));
  }

  getDraft(key: string): PublishRequest | undefined {
    const draft = this.drafts.get(key);
    return draft ? structuredClone(draft) : undefined;
  }

  deleteDraft(key: string): boolean {
    return this.drafts.delete(key);
  }
}

export class GatewayPublishClient {
  private readonly baseUrl: string;
  private readonly apiKey?: string;
  private readonly fetchImpl: typeof fetch;
  readonly draftCache = new LocalPublishDraftCache();

  constructor(options: GatewayPublishClientOptions) {
    this.baseUrl = options.baseUrl.replace(/\/+$/, "");
    this.apiKey = options.apiKey;
    this.fetchImpl = options.fetch ?? fetch;
  }

  async publish(request: PublishRequest): Promise<PublishedListing> {
    assertPublishRequestHasNoPayload(request);
    const response = await this.fetchImpl(`${this.baseUrl}/v1/gateway/publish`, {
      method: "POST",
      headers: this.headers(request.metadata),
      body: JSON.stringify(toWirePublishRequest(request)),
    });
    if (!response.ok) {
      throw new Error(`gateway publish failed: ${response.status} ${response.statusText}`);
    }
    return parsePublishedListing(await response.json());
  }

  async patchListing(listingId: OpaqueId, request: PatchListingRequest, ifMatch?: OpaqueId): Promise<PublishedListing> {
    assertNoPayloadDeep(request as unknown as Record<string, unknown>);
    const headers = this.headers(request.metadata);
    if (ifMatch) {
      headers["if-match"] = ifMatch;
    }
    const response = await this.fetchImpl(`${this.baseUrl}/v1/gateway/listings/${encodeURIComponent(listingId)}`, {
      method: "PATCH",
      headers,
      body: JSON.stringify(toWirePatchListingRequest(request)),
    });
    if (!response.ok) {
      throw new Error(`gateway listing patch failed: ${response.status} ${response.statusText}`);
    }
    return parsePublishedListing(await response.json());
  }

  private headers(metadata: SharedRequestMetadata): Record<string, string> {
    const headers: Record<string, string> = {
      "content-type": "application/json",
      "x-request-id": metadata.requestId,
    };
    if (metadata.idempotencyKey) {
      headers["idempotency-key"] = metadata.idempotencyKey;
    }
    if (this.apiKey) {
      headers.authorization = `Bearer ${this.apiKey}`;
    }
    return headers;
  }
}

export function assertPublishRequestHasNoPayload(request: PublishRequest): void {
  assertNoPayloadDeep(request as unknown as Record<string, unknown>);
}

function assertNoPayloadDeep(value: unknown): void {
  if (Array.isArray(value)) {
    value.forEach(assertNoPayloadDeep);
    return;
  }
  if (typeof value !== "object" || value === null) {
    return;
  }
  const record = value as Record<string, unknown>;
  assertNoPayloadFields(Object.keys(record));
  Object.values(record).forEach(assertNoPayloadDeep);
}

function toWirePublishRequest(request: PublishRequest): Record<string, unknown> {
  return dropUndefined({
    metadata: toWireMetadata(request.metadata),
    seller_id: request.sellerId,
    listing_metadata: toWireListingMetadata(request.listingMetadata),
    delivery_modes: request.deliveryModes,
    pricing: toWirePricing(request.pricing),
    sample_policy: request.samplePolicy ? toWireSamplePolicy(request.samplePolicy) : undefined,
    license_terms: toWireLicenseTerms(request.licenseTerms),
    trust_artifact_refs: request.trustArtifactRefs?.map(toWireSignedReference),
    seller_verification_ref: request.sellerVerificationRef ? toWireSignedReference(request.sellerVerificationRef) : undefined,
    listing_review_policy: request.listingReviewPolicy,
    abuse_signals: request.abuseSignals,
  });
}

function toWirePatchListingRequest(request: PatchListingRequest): Record<string, unknown> {
  return dropUndefined({
    metadata: toWireMetadata(request.metadata),
    listing_version_id: request.listingVersionId,
    listing_metadata: request.listingMetadata ? toWireListingMetadata(request.listingMetadata) : undefined,
    delivery_modes: request.deliveryModes,
    pricing: request.pricing ? toWirePricing(request.pricing) : undefined,
    sample_policy: request.samplePolicy ? toWireSamplePolicy(request.samplePolicy) : undefined,
    license_terms: request.licenseTerms ? toWireLicenseTerms(request.licenseTerms) : undefined,
    trust_artifact_refs: request.trustArtifactRefs?.map(toWireSignedReference),
    seller_verification_ref: request.sellerVerificationRef ? toWireSignedReference(request.sellerVerificationRef) : undefined,
    listing_review_policy: request.listingReviewPolicy,
    abuse_signals: request.abuseSignals,
  });
}

function toWireMetadata(metadata: SharedRequestMetadata): Record<string, unknown> {
  return dropUndefined({
    request_id: metadata.requestId,
    principal_id: metadata.principalId,
    account_id: metadata.accountId,
    traceparent: metadata.traceparent,
    idempotency_key: metadata.idempotencyKey,
    request_timestamp: metadata.requestTimestamp,
    nonce: metadata.nonce,
    signed_envelope: metadata.signedEnvelope,
  });
}

function toWireSignedReference(ref: SignedReference): Record<string, unknown> {
  return dropUndefined({
    ref: ref.ref,
    artifact_hash: ref.artifactHash,
    signer_identity: ref.signerIdentity,
    signature_state: ref.signatureState,
    verification_state: ref.verificationState,
    expires_at: ref.expiresAt,
  });
}

function toWireListingMetadata(metadata: ListingMetadata): Record<string, unknown> {
  return dropUndefined({
    title: metadata.title,
    summary: metadata.summary,
    tags: metadata.tags,
    category: metadata.category,
  });
}

function toWirePricing(pricing: ListingPricing): Record<string, unknown> {
  return {
    model: pricing.model,
    currency: pricing.currency,
    unit_amount_cents: pricing.unitAmountCents,
    billing_unit: pricing.billingUnit,
  };
}

function toWireSamplePolicy(policy: ListingSamplePolicy): Record<string, unknown> {
  return dropUndefined({
    preview_refs: policy.previewRefs?.map(toWireSignedReference),
    summary: policy.summary,
    row_count: policy.rowCount,
    column_names: policy.columnNames,
  });
}

function toWireLicenseTerms(terms: LicenseTerms): Record<string, unknown> {
  return dropUndefined({
    terms_ref: toWireSignedReference(terms.termsRef),
    commercial_use: terms.commercialUse,
    retention_days: terms.retentionDays,
  });
}

function parsePublishedListing(data: Record<string, unknown>): PublishedListing {
  return {
    listingId: pickString(data, "listing_id", "listingId"),
    listingVersionId: pickString(data, "listing_version_id", "listingVersionId"),
    sellerId: pickString(data, "seller_id", "sellerId"),
    state: pickString(data, "state") as PublishGovernanceState,
    discoverable: pickBoolean(data, "discoverable"),
    quoteEligible: pickBoolean(data, "quote_eligible", "quoteEligible"),
    listingMetadata: parseListingMetadata(pickRecord(data, "listing_metadata", "listingMetadata")),
    deliveryModes: pickStringArray(data, "delivery_modes", "deliveryModes"),
    pricing: parsePricing(pickRecord(data, "pricing")),
    samplePolicy: parseSamplePolicy(pickRecord(data, "sample_policy", "samplePolicy")),
    licenseTerms: parseLicenseTerms(pickRecord(data, "license_terms", "licenseTerms")),
    trustArtifactRefs: pickRecordArray(data, "trust_artifact_refs", "trustArtifactRefs").map(parseSignedReference),
    sellerVerificationRef: parseOptionalSignedReference(pickOptionalRecord(data, "seller_verification_ref", "sellerVerificationRef")),
    listingReviewPolicy: pickOptionalString(data, "listing_review_policy", "listingReviewPolicy"),
    governanceEvents: pickOptionalStringArray(data, "governance_events", "governanceEvents"),
    reasonCode: pickOptionalString(data, "reason_code", "reasonCode"),
  };
}

function parseSignedReference(data: Record<string, unknown>): SignedReference {
  return {
    ref: pickString(data, "ref"),
    artifactHash: pickOptionalString(data, "artifact_hash", "artifactHash"),
    signerIdentity: pickString(data, "signer_identity", "signerIdentity"),
    signatureState: pickString(data, "signature_state", "signatureState"),
    verificationState: pickString(data, "verification_state", "verificationState"),
    expiresAt: pickOptionalString(data, "expires_at", "expiresAt"),
  };
}

function parseOptionalSignedReference(data?: Record<string, unknown>): SignedReference | undefined {
  return data ? parseSignedReference(data) : undefined;
}

function parseListingMetadata(data: Record<string, unknown>): ListingMetadata {
  return {
    title: pickString(data, "title"),
    summary: pickString(data, "summary"),
    tags: pickOptionalStringArray(data, "tags"),
    category: pickOptionalString(data, "category"),
  };
}

function parsePricing(data: Record<string, unknown>): ListingPricing {
  return {
    model: pickString(data, "model"),
    currency: pickString(data, "currency"),
    unitAmountCents: pickNumber(data, "unit_amount_cents", "unitAmountCents"),
    billingUnit: pickString(data, "billing_unit", "billingUnit"),
  };
}

function parseSamplePolicy(data: Record<string, unknown>): ListingSamplePolicy {
  return {
    previewRefs: pickRecordArray(data, "preview_refs", "previewRefs").map(parseSignedReference),
    summary: pickOptionalString(data, "summary"),
    rowCount: pickOptionalNumber(data, "row_count", "rowCount"),
    columnNames: pickOptionalStringArray(data, "column_names", "columnNames"),
  };
}

function parseLicenseTerms(data: Record<string, unknown>): LicenseTerms {
  return {
    termsRef: parseSignedReference(pickRecord(data, "terms_ref", "termsRef")),
    commercialUse: pickOptionalBoolean(data, "commercial_use", "commercialUse"),
    retentionDays: pickOptionalNumber(data, "retention_days", "retentionDays"),
  };
}

function pickRecord(data: Record<string, unknown>, ...keys: string[]): Record<string, unknown> {
  const value = pick(data, ...keys);
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new Error(`expected object field ${keys[0]}`);
  }
  return value as Record<string, unknown>;
}

function pickOptionalRecord(data: Record<string, unknown>, ...keys: string[]): Record<string, unknown> | undefined {
  const value = pick(data, ...keys);
  if (value === undefined || value === null) {
    return undefined;
  }
  if (typeof value !== "object" || Array.isArray(value)) {
    throw new Error(`expected object field ${keys[0]}`);
  }
  return value as Record<string, unknown>;
}

function pickRecordArray(data: Record<string, unknown>, ...keys: string[]): Record<string, unknown>[] {
  const value = pick(data, ...keys);
  if (value === undefined) {
    return [];
  }
  if (!Array.isArray(value) || !value.every((item) => typeof item === "object" && item !== null && !Array.isArray(item))) {
    throw new Error(`expected object array field ${keys[0]}`);
  }
  return value as Record<string, unknown>[];
}

function pickString(data: Record<string, unknown>, ...keys: string[]): string {
  const value = pick(data, ...keys);
  if (typeof value !== "string" || value.length === 0) {
    throw new Error(`expected string field ${keys[0]}`);
  }
  return value;
}

function pickOptionalString(data: Record<string, unknown>, ...keys: string[]): string | undefined {
  const value = pick(data, ...keys);
  return typeof value === "string" ? value : undefined;
}

function pickNumber(data: Record<string, unknown>, ...keys: string[]): number {
  const value = pick(data, ...keys);
  if (typeof value !== "number") {
    throw new Error(`expected number field ${keys[0]}`);
  }
  return value;
}

function pickOptionalNumber(data: Record<string, unknown>, ...keys: string[]): number | undefined {
  const value = pick(data, ...keys);
  return typeof value === "number" ? value : undefined;
}

function pickBoolean(data: Record<string, unknown>, ...keys: string[]): boolean {
  const value = pick(data, ...keys);
  if (typeof value !== "boolean") {
    throw new Error(`expected boolean field ${keys[0]}`);
  }
  return value;
}

function pickOptionalBoolean(data: Record<string, unknown>, ...keys: string[]): boolean | undefined {
  const value = pick(data, ...keys);
  return typeof value === "boolean" ? value : undefined;
}

function pickStringArray(data: Record<string, unknown>, ...keys: string[]): string[] {
  const value = pick(data, ...keys);
  if (!Array.isArray(value) || !value.every((item) => typeof item === "string")) {
    throw new Error(`expected string array field ${keys[0]}`);
  }
  return value;
}

function pickOptionalStringArray(data: Record<string, unknown>, ...keys: string[]): string[] | undefined {
  const value = pick(data, ...keys);
  if (value === undefined) {
    return undefined;
  }
  if (!Array.isArray(value) || !value.every((item) => typeof item === "string")) {
    throw new Error(`expected string array field ${keys[0]}`);
  }
  return value;
}

function pick(data: Record<string, unknown>, ...keys: string[]): unknown {
  for (const key of keys) {
    if (key in data) {
      return data[key];
    }
  }
  return undefined;
}

function dropUndefined(data: Record<string, unknown>): Record<string, unknown> {
  return Object.fromEntries(Object.entries(data).filter(([, value]) => value !== undefined));
}
